libcudacxx/include/cuda/std/__atomic/functions/cuda_ptx_derived.h

```diff
@@ -339,6 +339,98 @@ _CCCL_DEVICE _Tp __atomic_exchange_n_cuda(_Tp volatile* __ptr, _Tp __val, int __
   return __ret;
 }
 
+template <typename _Tp, typename _Up, typename _Sco, __atomic_enable_if_not_native_arithmetic<_Tp> = 0>
+_CCCL_DEVICE _Tp __atomic_fetch_add_cuda(_Tp* __ptr, _Up __val, int __memorder, _Sco)
+{
+  return __atomic_fetch_update_cuda(
+    __ptr,
+    [__val](_Tp __old) {
+      return __old + __val;
+    },
+    __memorder,
+    _Sco{});
+}
+template <typename _Tp, typename _Up, typename _Sco, __atomic_enable_if_not_native_arithmetic<_Tp> = 0>
+_CCCL_DEVICE _Tp __atomic_fetch_add_cuda(volatile _Tp* __ptr, _Up __val, int __memorder, _Sco)
+{
+  return __atomic_fetch_update_cuda(
+    __ptr,
+    [__val](_Tp __old) {
+      return __old + __val;
+    },
+    __memorder,
+    _Sco{});
+}
+
+template <typename _Tp, typename _Up, typename _Sco, __atomic_enable_if_not_native_bitwise<_Tp> = 0>
+_CCCL_DEVICE _Tp __atomic_fetch_and_cuda(_Tp* __ptr, _Up __val, int __memorder, _Sco)
+{
+  return __atomic_fetch_update_cuda(
+    __ptr,
+    [__val](_Tp __old) {
+      return __old & __val;
+    },
+    __memorder,
+    _Sco{});
+}
+template <typename _Tp, typename _Up, typename _Sco, __atomic_enable_if_not_native_bitwise<_Tp> = 0>
+_CCCL_DEVICE _Tp __atomic_fetch_and_cuda(volatile _Tp* __ptr, _Up __val, int __memorder, _Sco)
+{
+  return __atomic_fetch_update_cuda(
+    __ptr,
+    [__val](_Tp __old) {
+      return __old & __val;
+    },
+    __memorder,
+    _Sco{});
+}
+
+template <typename _Tp, typename _Up, typename _Sco, __atomic_enable_if_not_native_bitwise<_Tp> = 0>
+_CCCL_DEVICE _Tp __atomic_fetch_or_cuda(_Tp* __ptr, _Up __val, int __memorder, _Sco)
+{
+  return __atomic_fetch_update_cuda(
+    __ptr,
+    [__val](_Tp __old) {
+      return __old | __val;
+    },
+    __memorder,
+    _Sco{});
+}
+template <typename _Tp, typename _Up, typename _Sco, __atomic_enable_if_not_native_bitwise<_Tp> = 0>
+_CCCL_DEVICE _Tp __atomic_fetch_or_cuda(volatile _Tp* __ptr, _Up __val, int __memorder, _Sco)
+{
+  return __atomic_fetch_update_cuda(
+    __ptr,
+    [__val](_Tp __old) {
+      return __old | __val;
+    },
+    __memorder,
+    _Sco{});
+}
+
+template <typename _Tp, typename _Up, typename _Sco, __atomic_enable_if_not_native_bitwise<_Tp> = 0>
+_CCCL_DEVICE _Tp __atomic_fetch_xor_cuda(_Tp* __ptr, _Up __val, int __memorder, _Sco)
+{
+  return __atomic_fetch_update_cuda(
+    __ptr,
+    [__val](_Tp __old) {
+      return __old ^ __val;
+    },
+    __memorder,
+    _Sco{});
+}
+template <typename _Tp, typename _Up, typename _Sco, __atomic_enable_if_not_native_bitwise<_Tp> = 0>
+_CCCL_DEVICE _Tp __atomic_fetch_xor_cuda(volatile _Tp* __ptr, _Up __val, int __memorder, _Sco)
+{
+  return __atomic_fetch_update_cuda(
+    __ptr,
+    [__val](_Tp __old) {
+      return __old ^ __val;
+    },
+    __memorder,
+    _Sco{});
+}
+
```

Comment:
TL;DR: There is no PTX for adding/manipulating 128b values atomically so they're all CAS loop emulated.